internal/crypto/ptls/config.go

```diff
@@ -0,0 +1,128 @@
+// Copyright 2024 the Pinniped contributors. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+package ptls
+
+import (
+	"crypto/tls"
+	"crypto/x509"
+	"slices"
+	"sync/atomic"
+
+	"k8s.io/apimachinery/pkg/util/sets"
+)
+
+// allowedCiphersForTLSOneDotTwo will only contain ciphers that meet the following criteria:
+// 1. They are secure
+// 2. They are returned by tls.CipherSuites
+// 3. They are returned by cipherSuitesForDefault or cipherSuitesForDefaultLDAP
+// This is atomic so that it can not be set and read at the same time.
+//
+//nolint:gochecknoglobals // This needs to be global because it will be set at application startup from configuration values
+var allowedCiphersForTLSOneDotTwo atomic.Value
+
+type SetAllowedCiphers func([]string) error
+
+// SetAllowedCiphersForTLSOneDotTwo allows configuration/setup components to constrain the allowed TLS ciphers for TLS1.2.
+// Not tested in unit tests for pollution reasons.
+func SetAllowedCiphersForTLSOneDotTwo(allowedCipherNames []string) error {
+	temp, err := validateAllowedCiphers(allowedCipherNames)
+	if err != nil {
+		allowedCiphersForTLSOneDotTwo.Store(temp)
+	}
+	return err
+}
+
+// getAllowedCiphersForTLSOneDotTwo returns the user-configured list of allowed ciphers for TLS1.2.
+// It is not exported so that it is only available to this package.
+func getAllowedCiphersForTLSOneDotTwo() []*tls.CipherSuite {
+	temp, ok := (allowedCiphersForTLSOneDotTwo.Load()).([]*tls.CipherSuite)
+	if ok {
+		return temp
+	}
+	return nil
+}
+
+// buildCipherSuites returns the intersection of its parameters, as a list of cipher suite IDs.
+// If allowedCipherSuites is empty, it will return all the configuredCipherSuites.
+func buildCipherSuites(
+	configuredCipherSuites []*tls.CipherSuite,
+	allowedCipherSuites []*tls.CipherSuite,
+) []uint16 {
+	allowedCipherSuiteIDs := sets.New[uint16]()
```

Comment:
Would this work as a little performance optimization? All this intersecting and sorting seems expensive when you already know the answer in the case of the default configuration.

```suggestion
	if len(allowedCipherSuites) == 0 {
		return configuredCipherSuites
	}
	allowedCipherSuiteIDs := sets.New[uint16]()
```